src/tls/context.h

```diff
@@ -2,8 +2,518 @@
 // Licensed under the Apache 2.0 License.
 #pragma once
 
-#ifdef TLS_PROVIDER_IS_MBEDTLS
-#  include "mbedtls/context.h"
-#else
-#  include "openssl/context.h"
-#endif
+#include "cert.h"
+#include "crypto/base64.h"
+#include "crypto/entropy.h"
+#include "ds/logger.h"
+#include "tls/tls.h"
+
+#include <memory>
+#include <openssl/bio.h>
+#include <openssl/ssl.h>
+
+using namespace crypto;
+
+namespace tls
+{
+  class Context
+  {
+  protected:
+    crypto::OpenSSL::Unique_SSL_CTX cfg;
+    crypto::OpenSSL::Unique_SSL ssl;
+
+  public:
+    Context(bool client, bool dtls) :
+      cfg(
+        dtls ? (client ? DTLS_client_method() : DTLS_server_method()) :
+               (client ? TLS_client_method() : TLS_server_method())),
+      ssl(cfg)
+    {
+      // Require at least TLS 1.2, support up to 1.3
+      SSL_CTX_set_min_proto_version(
+        cfg, dtls ? DTLS1_2_VERSION : TLS1_2_VERSION);
+      SSL_set_min_proto_version(ssl, dtls ? DTLS1_2_VERSION : TLS1_2_VERSION);
+
+      // Disable renegotiation to avoid DoS
+      SSL_CTX_set_options(
+        cfg,
+        SSL_OP_CIPHER_SERVER_PREFERENCE |
+          SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION |
+          SSL_OP_NO_RENEGOTIATION);
+      SSL_set_options(
+        ssl,
+        SSL_OP_CIPHER_SERVER_PREFERENCE |
+          SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION |
+          SSL_OP_NO_RENEGOTIATION);
+
+      // Set cipher for TLS 1.2 (TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256)
+      SSL_CTX_set_cipher_list(
+        cfg,
+        "ECDHE-ECDSA-AES256-GCM-SHA384:"
+        "ECDHE-ECDSA-AES128-GCM-SHA256");
+      SSL_set_cipher_list(
+        ssl,
+        "ECDHE-ECDSA-AES256-GCM-SHA384:"
+        "ECDHE-ECDSA-AES128-GCM-SHA256");
+
+      // Set cipher for TLS 1.3 (same as above)
+      SSL_CTX_set_ciphersuites(
+        cfg,
+        "TLS_AES_256_GCM_SHA384:"
+        "TLS_AES_128_GCM_SHA256");
+      SSL_set_ciphersuites(
+        ssl,
+        "TLS_AES_256_GCM_SHA384:"
+        "TLS_AES_128_GCM_SHA256");
```

Comment:
So we no longer support the restrict cipherlist flag in cmake, which is fine, can you delete it too please (and its mention in the doc)?